### PR TITLE
sg: check dev-private rather than forced update

### PR DESCRIPTION
**Pull request by burmudar**
Based on the discussion [here](https://github.com/sourcegraph/sourcegraph/pull/42531#issuecomment-1268447884) I thought I take @mrnugget suggestion for a ride and it seems nice 👌🏼 

### An error occured
Screenshot 2022-10-06 at 11 15 59

### You're missing out on changes
Screenshot 2022-10-06 at 11 16 41

### No changes
Screenshot 2022-10-06 at 11 17 57

## Test plan
Tested locally

### CODE REVIEW

File: dev/sg/sg_start.go [hunk 1]
```diff
@@ -214,25 +214,40 @@ func startExec(ctx *cli.Context) error {
...
 	return startCommandSet(ctx.Context, set, config)
 }
 
+func shouldUpdateDevPrivate(ctx context.Context, path string) (bool, error) {
+	// git fetch so that we check whether there are any remote changes
+	if err := sgrun.Bash(ctx, "git fetch origin master").Dir(path).Run().Wait(); err != nil {
+		return false, err
+	}
+	// now we check if there are any changes ie. if the output is empty, we're not missing out on anything
+	output, err := sgrun.Bash(ctx, "git diff --shortstat origin/HEAD").Dir(path).Run().String()
```
Comment (mrnugget) [inline dev/sg/sg_start.go:243]: Why `origin/HEAD` here and not `origin/master`?

File: dev/sg/sg_start.go [hunk 2]
```diff
@@ -214,25 +214,40 @@ func startExec(ctx *cli.Context) error {
...
 
 	return startCommandSet(ctx.Context, set, config)
 }
 
+func shouldUpdateDevPrivate(ctx context.Context, path string) (bool, error) {
+	// git fetch so that we check whether there are any remote changes
+	if err := sgrun.Bash(ctx, "git fetch origin master").Dir(path).Run().Wait(); err != nil {
+		return false, err
+	}
+	// now we check if there are any changes ie. if the output is empty, we're not missing out on anything
```
Comment (mrnugget) [inline dev/sg/sg_start.go:242]: Nit:

```suggestion
	// Now we check if there are any changes. If the output is empty, we're not missing out on anything.
```

File: dev/sg/sg_start.go [hunk 3]
```diff
@@ -214,25 +214,40 @@ func startExec(ctx *cli.Context) error {
...
-		update := std.Out.Pending(output.Styled(output.StylePending, "Updating dev-private..."))
-		if err := sgrun.Bash(ctx.Context, "git pull origin master").
-			Dir(devPrivatePath).
-			Run().Wait(); err != nil {
-
+		// dev-private exists, let's see if there are any changes
+		update := std.Out.Pending(output.Styled(output.StylePending, "Checking for dev-private changes..."))
+		shouldUpdate, err := shouldUpdateDevPrivate(ctx.Context, devPrivatePath)
+		if shouldUpdate {
+			update.WriteLine(output.Line(output.EmojiInfo, output.StyleSuggestion, "We found some changes in dev-private that you're missing out on! If you want the new changes, do a git stash and then a git pull!"))
```
Comment (mrnugget) [inline dev/sg/sg_start.go:221]:
```suggestion
			update.WriteLine(output.Line(output.EmojiInfo, output.StyleSuggestion, "We found some changes in dev-private that you're missing out on! If you want the new changes, 'cd ../dev-private' and then do a 'git stash' and a 'git pull'!"))
```

**Review by mrnugget — APPROVED** [review]
[in Borat voice]: Very nice!

File: dev/sg/sg_start.go (hunk 1 above)
Comment (burmudar) [reply dev/sg/sg_start.go:243]: Figured HEAD would be slightly future proof incase we switch the default branch in dev-private to `main` 🤔

Comment (mrnugget) [reply dev/sg/sg_start.go:243]: Ah, I see. Well, then let me be super nitpicky and say: shouldn't we then pull out the name into a constant and use that in both places?

```go
const devPrivateDefaultBranch = "master"
// [...]
sgrun.Bash(ctx, fmt.Sprintf("git fetch origin %s", devPrivateDefaultBranch))
// [...]
sgrun.Bash(ctx, fmt.Sprintf("git diff --shortstat origin/%s", devPrivateDefaultBranch))
```

Comment (burmudar) [reply dev/sg/sg_start.go:243]: Ah yeah that is nice. I'll update it 👍🏼

**Pull request merged**